src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx

```diff
@@ -197,61 +421,141 @@ template <typename TargetT>
-  //  list_mode_data_sptr->set_get_position(start_time);
-  // TODO implement function that will do this for a random time
-  this->list_mode_data_sptr->reset();
-  double current_time = 0.;
-  ProjMatrixElemsForOneBin proj_matrix_row; 
-  shared_ptr<CListRecord> record_sptr = this->list_mode_data_sptr->get_empty_record_sptr(); 
-  CListRecord& record = *record_sptr; 
-  //  int count_of_events=0;
-  //int in_the_range =0;
-  while (this->list_mode_data_sptr->get_next_record(record) == Succeeded::yes) 
-  { 
-    //count_of_events++;
-    if(record.is_time())
-      {
-        current_time = record.time().get_time_in_secs();
-      }
-    if (current_time >= end_time)
-      {
-        break; // get out of while loop
-      }
-    if (current_time < start_time)
-      continue;
-    if (record.is_event() && record.event().is_prompt()) 
-      { 
-        Bin measured_bin; 
-        record.event().get_bin(measured_bin, *proj_data_info_cyl_uncompressed_ptr); 
-        if (measured_bin.get_bin_value() <= 0)
-          continue;      
-        this->PM_sptr->get_proj_matrix_elems_for_one_bin(proj_matrix_row, measured_bin); 
-        //in_the_range++;
-        Bin fwd_bin; 
-        proj_matrix_row.forward_project(fwd_bin,current_estimate); 
-        // additive sinogram 
-        if (!is_null_ptr(this->additive_proj_data_sptr))
-          {
-            float add_value = this->additive_proj_data_sptr->get_bin_value(measured_bin);
-            float value= fwd_bin.get_bin_value()+add_value;         
-            fwd_bin.set_bin_value(value);
-          }
-        float  measured_div_fwd = measured_bin.get_bin_value()/fwd_bin.get_bin_value();
-        measured_bin.set_bin_value(measured_div_fwd);
-        proj_matrix_row.back_project(gradient, measured_bin); 
-         
-      } 
-  }
-  //  cerr << " The number_of_events " << count_of_events << "   ";
-  //cerr << " The number of events proecessed "  << in_the_range << "  ";
-    
+    assert(subset_num>=0);
+    assert(subset_num<this->num_subsets);
+
+    ProjDataInfoCylindricalNoArcCorr* proj_data_no_arc_ptr =
+            dynamic_cast<ProjDataInfoCylindricalNoArcCorr *> (proj_data_info_cyl_sptr.get());
+
+    const double start_time = this->frame_defs.get_start_time(this->current_frame_num);
+    const double end_time = this->frame_defs.get_end_time(this->current_frame_num);
+
+    long num_used_events = 0;
+    const float max_quotient = 10000.F;
+
+    //go to the beginning of this frame
+    //  list_mode_data_sptr->set_get_position(start_time);
+    // TODO implement function that will do this for a random time
+    this->list_mode_data_sptr->reset();
+    double current_time = 0.;
+    ProjMatrixElemsForOneBin proj_matrix_row;
+
+    shared_ptr<CListRecord> record_sptr = this->list_mode_data_sptr->get_empty_record_sptr();
+    CListRecord& record = *record_sptr;
+
+    // Daniel: look into all the frames defined into the fdef file
+    for (this->current_frame_num = 1;
```

Comment:
@danieldeidda I don't understand this loop. 
are we reconstructing a single image here from several time frames? That seems rather non-standard. I'd expect here to be only 1 time frame (and that would need to be checked).

Reply:
I agree with @KrisThielemans. I will see where this code fits better

Reply:
Hi,
Sorry I missed Kris comment. I had done this loop because whenever I wanted to reconstruct short frames that were not at the beginning of the LM file reconstruction was not working. For example:

0 t0
0 t1
1 t1

Reply:
if the other frames have the zero before:
0 t0
these will not be taken into account. Another reason for this could be that one could want to reconstruct  pieces of data which are not consecutive (maybe there is strong motion or something bad) 
1 t0
0 t1
1 t2.
With this loop it worked for all of my data.

Reply:
I'm still confused. The first case (initial zeroes) is essential to cover, but doesn't require a loop over time frames (there will be only 1). The second case in my opinion is normally only useful if each time frame is reconstructed as a different image (as you'll have motion between them).
 
Another case would be when the data is gated. Then your example would be useful (i.e. every "interval"  belongs to the same gate). However, then we end up mixing gating and TimeFrameDefinitions, which sounds like a terrible idea (even though it'd be handy). ok, I did it for the (non-public) `lm_to_gated_projdata` but it's one reason I didn't want to release that yet :-; To handle this properly, we'd need to be able to have (dynamic) gated data, and an appropriate data structure for that. That is surely another PR.

Summary, I'd rather not have this loop (and add a check that `num_time_frames==1`).